### PR TITLE
Fix Parameter config class

### PR DESCRIPTION
**Pull request by cebrusfs**
For issue #398.

For 1, Not sure why the format string will not work in origin environment. Since I tested the class which support `__str__` can work with `%s` like the following.
```
class A:
    def __init__(self):
        self.value = 'meow'
    def __str__(self):
        return str(self.value)
a = A()
print('%s' % a)
```
Currently it works perfectly now, so I won't implement `__format__`.

For 2, see fcab5d3.
I remap the member name using `__setattr__` and `__getattribute__`. The gdb python implementation will let any value from set command write to `gdb.Parameter.value`. However, I remap the member name of Parameter by the following rule
* `Parameter.value` will map to `gdb.Parameter._value` (if it is string type, always keep unicode)
* `Parameter.raw_value` will map to `gdb.Parameter.value`

Also,
* Fix `.split()`
* Add `revert_default()`. Then people can just write `param.revert_default()` instead of `param.value = param.default` 
* Fix operator overwriting of `Parameter`
   * Rewrite `__lt__` and `__eq__`

### CODE REVIEW

File: pwndbg/config.py [hunk 1]
```diff
@@ -86,7 +88,24 @@ def value_to_gdb_native(value):
...
 
 
+member_remap = {
+    'value': '_value',
+    'raw_value': 'value'
+}
+@total_ordering
 class Parameter(gdb.Parameter):
+    """
+    For python2, we can not store unicode type in self.value since the implemetation limitation of gdb python.
```
Comment (disconnect3d) [inline pwndbg/config.py]: Typo `implemetation` -> `implementation`

File: pwndbg/config.py [hunk 2]
```diff
@@ -86,7 +88,24 @@ def value_to_gdb_native(value):
...
+    We use self._value as the converted cache and set __getattribute__() and __setattr__() to remaping the variable.
```
Comment (disconnect3d) [inline pwndbg/config.py]: `to remap variables`

File: pwndbg/config.py [hunk 3]
```diff
@@ -86,7 +88,24 @@ def value_to_gdb_native(value):
...
+
+    Since gdb python will set gdb.Parameter.value as user input and call get_set_string() as setter, we use self.raw_value to map back to gdb.Parameter.value
```
Comment (disconnect3d) [inline pwndbg/config.py]:
```
Since GDB will set gdb.Parameter.value to user input and call get_set_string() ` (...)

File: pwndbg/config.py [hunk 4]
```diff
@@ -135,30 +184,56 @@ def __str__(self):
     def __bool__(self):
         return bool(self.value)
 
+    # Comparing operaters
```
Comment (disconnect3d) [inline pwndbg/config.py]: `Compare operators`

File: pwndbg/config.py [hunk 5]
```diff
@@ -135,30 +184,56 @@ def __str__(self):
...
     def __lt__(self, other):
-        return self.optname <= other.optname
+        if isinstance(other, gdb.Parameter):
+            return self.optname < other.optname
+        else:
+            return self.value < other
 
-    def __div__(self, other):
-        return self.value / other
+    # Operaters
```
Comment (disconnect3d) [inline pwndbg/config.py]: `Operators`

**Comment by disconnect3d** [comment]
@anthraxx u are more into the topic; is this design okay? can we push it forward and merge it? or we want to make those things totally differently?

@zachriggle if u have time, pls visit this too

@cebrusfs btw this has a conflict, rebase to current dev pls

File: pwndbg/config.py [hunk 6]
```diff
@@ -135,30 +184,56 @@ def __str__(self):
...
+    def __add__(self, other):
+        return self.value + other
 
-    def __floordiv__(self, other):
-        return self.value // other
+    def __radd__(self, other):
```
Comment (zachriggle) [inline pwndbg/config.py:213]: I feel like there is some weird abstract base class we can inherit from that will automatically add these math operators.  I'd prefer using that rather than specifying each individual method. 

If there's not one, we should create one.

File: pwndbg/config.py [hunk 7]
```diff
@@ -135,30 +184,56 @@ def __str__(self):
...
+    def __eq__(self, other):
+        if isinstance(other, gdb.Parameter):
+            return self.optname == other.optname
+        else:
+            return self.value == other
+
     def __lt__(self, other):
-        return self.optname <= other.optname
+        if isinstance(other, gdb.Parameter):
+            return self.optname < other.optname
```
Comment (zachriggle) [inline pwndbg/config.py:205]: Why are we comparing the name on `<`?

File: pwndbg/config.py [hunk 8]
```diff
@@ -113,19 +132,49 @@ def native_default(self):
...
+        return super(Parameter, self).__getattribute__(new_name)
+
     def get_set_string(self):
+        value = self.raw_value
+
+        # For string value, convert utf8 byte string to unicode.
+        if isinstance(value, six.binary_type):
+            value = codecs.decode(value, 'utf-8')
+
+        # Remove surrounded ' and " characters
```
Comment (zachriggle) [inline pwndbg/config.py:154]: I'm not sure that stripping quotes is a good idea. 

* Where are quotes necessary?
* What existing cases do we use quotes?
* Doesn't GDB auto-strip quotes?
* How do I intentionally force quotes into a value?

Comment (cebrusfs) [reply pwndbg/config.py:154]: The original code implement this and use `replace` to do that. I think it is because `gdb.Parametor` will keep all characters. That is, `set color 'bold'` will get a string `'bold'` rather than `bold`. I guess people this to make sure  `set color 'bold'` works the same as `set color bold` since programmer often like to add quotes on string literal.

> How do I intentionally force quotes into a value?

Currently we can't put quote at the beginning and ending. Before my commit, your implement use `replace` to do that. It will remove all the quotes in the string, but I don't see any complain about that. Use `strip` will much better than before.

File: pwndbg/config.py (hunk 7 above)
Comment (cebrusfs) [reply pwndbg/config.py:205]: It is used to display all config in `config()` of `command/config.py`. It put all parameter in a list and then sort.

File: pwndbg/config.py (hunk 6 above)
Comment (cebrusfs) [reply pwndbg/config.py:213]: I don't find a good base class to do that. Our use case is weird because we just want to proxy all operator to `self.value`. People often inherit the type of `value` to do that in this use case. However, we can't do that because the type of `value` is dynamic and reply on gdb's implementation. That's why we need a operator proxy to wrap the operators.

I think this use case is limited and even we create a abstract class for this, it will be only one class will use it. At least, until now we only have one class that need to implement operator. What do you think?

File: pwndbg/config.py (hunk 8 above)
Comment (cebrusfs) [reply pwndbg/config.py:154]: @anthraxx @disconnect3d 
BTW, I saw the PR #409 reply on the quotes. However, the implementation of `dev` `HEAD` will remove all quotes by `replace`. 
https://github.com/pwndbg/pwndbg/blob/103b1def610957e2c74ab616b15a0513a1286cd6/pwndbg/config.py#L120

**Comment by cebrusfs** [comment]
Fix all @disconnect3d mentioned typos and thanks for that.

Also, rebased to `dev`.

File: pwndbg/config.py (hunk 8 above)
Comment (cebrusfs) [reply pwndbg/config.py:154]: It seems it will be set after `trigger()` so the checker will get the raw value. So PR #409 works fine and will be broken by this PR.

However, the in-consistence between trigger and final value will be a problem? For example, the trigger function does not expect the quote will be removed.

Can we re-design what we expect to do before/after the trigger function?

**Comment by zachriggle** [comment]
As much as I hate eval(), it may be worth attempting to eval() the values
using something like Pwntools’ safe_eval rather than stripping quotes.

It’s a relatively small chunk of code that might be useful anyway, and
should have more predictable results than just quote-stripping.

**Comment by cebrusfs** [comment]
Sound better than `strip`. But we have more problems here.
* When should we do this kind of `eval` or casting? Since there may be some wrapper function called trigger function which may modify the value. So what kind of value should the wrapper get and return? with quote or without quote?
* if the value is without quote, will there be any problem that using safe_eval. Or we need some error handling on this case. Currently implementation looks it want to accept both kind of value.

**Comment by zachriggle** [comment]
The idea would be that failure to eval would just use the raw string

File: pwndbg/config.py (hunk 8 above)
Comment (disconnect3d) [reply pwndbg/config.py:154]: @cebrusfs I like the idea of removing (well replacing) the quotes.

I don't really understand why would it break #409? Anyway, if so, it shouldn't :P

**Comment by anthraxx** [comment]
So far I don't think it will break, but as I heard explicitly "stripping" , please keep in mind that strings containing only whitespace are important to be supported, like putting them in the quotes to work.
They are used for style relevant alignment aka theme. So a prefix configuration value must be able to hold like 2 space characters without any printable character.

**Comment by cebrusfs** [comment]
@anthraxx Yep, it is supported. My implementation only strip the quotes (`'` and `"`).

**Comment by disconnect3d** [comment]
@cebrusfs mm and what if someone wants to use `"` inside a string config variable? :P

File: pwndbg/config.py (hunk 8 above)
Comment (cebrusfs) [reply pwndbg/config.py:154]: @disconnect3d 
The original implementation has these feature. I just use `strip` to remove quotes instead of use `replace` to avoid break some string that contains quotes in the middle. Ex. `"me'ow"` should be `me'ow` (after), not `meow` (before).

I said it will break since the quote removing happened after calling wrapper function (`Trigger`) in the original version.  That is the value `'meow'` will still be `'meow'` when the wrapper fetches the value from `self.value` and after that it will be striped and becomes `meow`.

However, my implementation removes the quotes at the first, so the wrapper will get `meow` in these case. So `""` and `''` case will not happend in your code `config_context_sections.value.lower() in ('""', "''", 'none', 'empty'):`.

One way to fix that is moving the quote replacing step after calling the wrapper function (keep original order).

However, this preprocess happens after calling the wrapper functions is weird since the needed of removing quote is caused by gdb argument parser. In python environment (wrapper function), I think it is better to not to have any extra quote striping.

**Comment by cebrusfs** [comment]
@disconnect3d For now, I don't support to set the value as a single quote but it is possible to put the quote between some no-quote characters. However, before my PR, it uses `replace` to do that, so any quote in the string will be removed.  I don't see any issue mentioned his need to put any quote in the config variable.

Maybe we can think how to parse it correctly like use regex to replace `'(.+)'` and `"(.+)"`? For me, I prefer a weak workaround because I don't like eval based solution, it is overkill.

**Comment by zachriggle** [comment]
Can we split the quote-stripping-or-eval-or-whatever stuff into a separate pull request?

**Comment by cebrusfs** [comment]
@zachriggle @disconnect3d @anthraxx 
Hi all,

I change to use regex to remove the quote. Generally, we can put any string in parameter now, but does not support escape sequence. If someone need this, you can create the other PR.
You can set the parameter to `''` by inputting `"''"` or `''''`.

I keep the "quote removing" step before "calling the wrapper function". I review all wrapper function and no one will be affect expect `context` command in PR #409.
However, I found out it will NOT break PR #409 (sorry) since the empty string is handled by the first  condition `not config_context_sections.value`. The only problem is two cases will not useful `""` and `''` since we already removed that quotes. So, I removed them.

If you don't have any concern, you can merge now.

BTW, For the regex, I have some tests on regex pattern, you can add it to autotests after you finished the test framework.
```python
import re

# The first character must be ' or " and ends with the same character.
pattern = r"^(?P<quote>[\"'])(?P<content>.*?)(?P=quote)$"

tests = [
    ("''", ""),
    ('""', ""),

    ("'a'", "a"),
    ('"a"', "a"),

    ("'''", "'"),
    ('"""', '"'),

    ('"\'', None),
    ('\'"', None),

    ('""\'', None),
    ('\'\'"', None),
    ('"\'\'', None),
    ('\'""', None),

    ("a'bc", None),
    ("a\"bc", None),

    ("'a\"b\'c'", "a\"b\'c"),
    ("'a\'b\"c'", "a\'b\"c"),
    ('"a\"b\'c"', "a\"b\'c"),
    ('"a\'b\"c"', "a\'b\"c"),
]


for inp, ans in tests:
    out = re.sub(pattern, r"\g<content>", inp)

    if ans is None: ans = inp

    print repr(inp), repr(out), repr(ans)
    assert out == ans

print 'passed'
```

**Review by disconnect3d — APPROVED** [review]
lgtm. Seems to be working fine for both Py2 and Py3.

@zachriggle and @anthraxx can u approve or request changes? I am not going to merge without at least @anthraxx approval here :D (and will try to give some time for @zachriggle to have chance to approve/reject :P).

**Review by zachriggle — APPROVED** [review]
Not really sure about the quoting stuff, but the rest is fine and the quoting doesn't affect me directly so #yolo

**Comment by disconnect3d** [comment]
@cebrusfs btw this needs rebase and fixing conflicts :(.

@anthraxx bump - review pls! :P

**Comment by cebrusfs** [comment]
@disconnect3d 
Rebased and reverted #416 since the Parameter object did not have `.decode()`, and it will broken this PR.

**Comment by anthraxx** [comment]
@cebrusfs w00t, why you revert a different fix because it breaks your PR? It the formulation wrong or is the approach reverting is wrong? The problem that commit fixed should be preserved and not reverted.

on py3 system all looks okayish on read and rought testing but im not really sure about py2 and multichar padding after that revert, needs explicit testing with a py2 env if those funcs _always_ get unicode instances

**Comment by cebrusfs** [comment]
@anthraxx sorry for misleading.  The main reason is not that it breaks my PR. It is because #416 is a work around for some unicode character used in config for py2 and my PR can completely fix this. The comment also mentioned it is a workaround until this PR has pushed.
The other problem is, in the current design, the `x` passed in the ljust, rjust functions is not string type. It is `Parameter` class because Parameter class is designed for wrapper/proxy class of the `value`. It supports most of operators, but does not support to call the un-implement member function such as `decode()`.
We have two way to fix this.
1. add implementation or remove the using of decode
But we will not need #416 anymore and have plan to remove it, so I decide to revert it.
2. pass all unknown member function calling of `Parameter` to the `value`'s type.
I think it is a better way to do that, but it can be future work.

For the tests, I tested on the py2 env and works fine. The original issue #412 is caused by the config `Parameter` class provide the utf8 sequence but in this PR we change to keep everything is `unicode` type.

**Comment by anthraxx** [comment]
sounds good, i guess it should all work then.
